THORN-2051: MP metrics: MetricRegistry.getTimers and similar methods don't filter by type

Ladicek opened the pull request:
Motivation
----------
MP Metrics allow to get all registered metrics of given type
using `MetricRegistry.getTimers()` (and similar: `getMeters`,
`getCounters` etc.). However, these methods currently always
return all registered metrics of all types. That's clearly
wrong and actually ultimately results in `ClassCastException`s.

Modifications
-------------
The internal method `MetricsRegistryImpl.getMetrics`,
which is ultimately called in all these cases, already takes
a `MetricType`. It just wasn't used. Modified the loop to,
in addition to checking the filter, also test the metric type.

Result
------
`MetricRegistry.getTimers` and similar methods now work correctly.

- [x] Have you followed the guidelines in our [Contributing](http://wildfly-swarm.io/community/contributing/) document?
- [x] Have you created a [JIRA](https://issues.jboss.org/browse/THORN) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/thorntail/thorntail/pulls) for the same issue?
- [ ] Have you built the project locally prior to submission with `mvn clean install`?

-----

Ladicek committed:
THORN-2051: MP metrics: MetricRegistry.getTimers and similar methods don't filter by type

Motivation
----------
MP Metrics allow to get all registered metrics of given type
using `MetricRegistry.getTimers()` (and similar: `getMeters`,
`getCounters` etc.). However, these methods currently always
return all registered metrics of all types. That's clearly
wrong and actually ultimately results in `ClassCastException`s.

Modifications
-------------
The internal method `MetricsRegistryImpl.getMetrics`,
which is ultimately called in all these cases, already takes
a `MetricType`. It just wasn't used. Modified the loop to,
in addition to checking the filter, also test the metric type.

Result
------
`MetricRegistry.getTimers` and similar methods now work correctly.

mkouba commented on fractions/microprofile/microprofile-metrics/src/main/java/org/wildfly/swarm/microprofile/metrics/runtime/MetricsRegistryImpl.java:
```diff
@@ -286,13 +286,15 @@ public void removeMatching(MetricFilter metricFilter) {
             if (filter.matches(entry.getKey(), entry.getValue())) {
-                out.put(entry.getKey(), (T) entry.getValue());
+                Metadata metadata = metadataMap.get(entry.getKey());
+                assert metadata != null : "missing metadata for registered metric";
```
So if metadata are missing and assertions are disabled the next line would just throw NPE?

Ladicek replied:
See https://github.com/Ladicek/wildfly-swarm/blob/5ac9db2cee25adc0a44d51e753e40458692dd2a8/fractions/microprofile/microprofile-metrics/src/main/java/org/wildfly/swarm/microprofile/metrics/runtime/MetricsRegistryImpl.java#L82-L84 and https://github.com/Ladicek/wildfly-swarm/blob/5ac9db2cee25adc0a44d51e753e40458692dd2a8/fractions/microprofile/microprofile-metrics/src/main/java/org/wildfly/swarm/microprofile/metrics/runtime/MetricsRegistryImpl.java#L120-L121

mkouba commented:
Is this use case tested by tck? Shouldn't we add our own test?

mkouba commented:
By the way, there is no thorntail-specific test suite for metrics yet. But I have a pom prepared and I'm going to include this in PR for [THORN-2050](https://issues.jboss.org/browse/THORN-2050).

Ladicek commented:
I added a test in https://github.com/Ladicek/swarm-test-suite/commit/f7e4e3256f3c803468658de3b1bfc100390e66a5

Ladicek commented:
Rebased and changed order of tests: first test if metric is of expected type, then test if it matches the filter. This is undefined by MP Metrics (filed https://github.com/eclipse/microprofile-metrics/issues/248), but I believe that it makes more sense like this -- if someone calls `MetricRegistry.getTimers(myFilter)`, they can easily assume that `myFilter` will only be passed timers.

Also added a test for the fix.

Ladicek commented:
WAT?

retest this please

Ladicek commented:
Windows build failed with the 2 known MP FT TCK failures.
